src/ComputeSharp.D2D1.CodeFixers/MissingPixelShaderDescriptorOnPixelShaderCodeFixer.cs

```diff
@@ -0,0 +1,103 @@
+    {
+        return WellKnownFixAllProviders.BatchFixer;
+    }
+
+    /// <inheritdoc/>
+    public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+    {
+        Diagnostic diagnostic = context.Diagnostics[0];
+        TextSpan diagnosticSpan = context.Span;
+
+        SyntaxNode? root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+        // Get the struct declaration from the target diagnostic
+        if (root!.FindNode(diagnosticSpan) is StructDeclarationSyntax structDeclaration)
+        {
+            // Register the code fix to update the return type to be Task instead
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    title: "Add [D2DGeneratedPixelShaderDescriptor] attribute",
+                    createChangedDocument: token => ChangeReturnType(context.Document, root, structDeclaration),
+                    equivalenceKey: "Add [D2DGeneratedPixelShaderDescriptor] attribute"),
+                diagnostic);
+        }
+    }
+
+    /// <summary>
+    /// Applies the code fix to add the [D2DGeneratedPixelShaderDescriptor] attribute to a target type.
+    /// </summary>
+    /// <param name="document">The original document being fixed.</param>
+    /// <param name="root">The original tree root belonging to the current document.</param>
+    /// <param name="structDeclaration">The <see cref="StructDeclarationSyntax"/> to update.</param>
+    /// <returns>An updated document with the applied code fix, and the return type of the method being <see cref="Task"/>.</returns>
+    private static Task<Document> ChangeReturnType(Document document, SyntaxNode root, StructDeclarationSyntax structDeclaration)
+    {
+        int index = 0;
+
+        // Find the index to use to insert the attribute. We want to make it so that if the struct declaration
+        // has a bunch of D2D attributes, the new one will be inserted right after that. This way the final list
+        // will be nicely sorted, instead of having D2D attributes interleaving other unrelated attributes, if any.
+        foreach (AttributeListSyntax attributeList in structDeclaration.AttributeLists)
+        {
+            // Make sure we have an attribute to check
+            if (attributeList.Attributes is not [AttributeSyntax attribute, ..])
+            {
+                continue;
+            }
+
+            // Make sure we find a readable identifier for the attribute
+            if (attribute.Name is not IdentifierNameSyntax { Identifier.Value: string identifier })
+            {
+                break;
+            }
+
+            // If the attribute is D2D one, increment the index and continue
+            if (identifier.Contains("D2D"))
+            {
+                index++;
+            }
+            else
+            {
+                // Otherwise, stop here, we reached the end of the sequence
+                break;
+            }
```

Comment:
This is questionable. It only checks the first `AttributeSyntax` from `attributeList.Attributes`.

Reply:
This is by design, as in, I want to insert the new attribute list as the first one after the ones with D2D attributes, if any. I'm assuming that even if you have multiple attributes if one of those lists, if the first one is a D2D one, the following ones would also be (as you might group them by semantics). So I just want to find the index of the last such attribute list. Does that make sense?

Also, do you reckon I should be checking the attributes by symbol instead of just trying to match the identifier name? Ie. so that it'd also handle eg. type aliases? 🤔

Reply:
I mean, if the attribute list is:

`[UnrelatedAttribute, D2D_Something]`

you will insert at index 0 (if I'm reading the code correctly)

Reply:
Ah, right. Yeah I need to switch this to check symbols 😅

Reply:
Fixed in 112a4af860d6082868d122e7ef308cd6c9c56014, thank you! 😄